core/src/main/java/org/infinispan/expiration/impl/TxClusterExpirationManager.java

```diff
@@ -12,11 +19,46 @@
 
    @Inject
    protected KeyPartitioner partitioner;
+   @Inject @ComponentName(KnownComponentNames.PERSISTENCE_EXECUTOR)
+   protected ExecutorService blockingExecutor;
+
+   private boolean optimisticTransaction;
+
+   @Override
+   public void start() {
+      super.start();
+
+      optimisticTransaction = configuration.transaction().lockingMode() == LockingMode.OPTIMISTIC;
+   }
 
    @Override
    CompletableFuture<Boolean> handleMaxIdleExpireEntry(K key, V value, long maxIdle, boolean skipLocking) {
       // We cannot remove an entry in a tx, due to possible dead lock - thus we just return
       // if the entry is expired or not
-      return retrieveLastAccess(key, value, partitioner.getSegment(key)).thenApply(hasMaxIdle -> hasMaxIdle == null);
+      return retrieveLastAccess(key, value, partitioner.getSegment(key)).thenApply(Objects::isNull);
+   }
+
+   @Override
+   CompletableFuture<Void> removeLifespan(AdvancedCache<K, V> cacheToUse, K key, V value, long lifespan) {
+      // Transactional is still blocking - to be removed later
+      return CompletableFuture.supplyAsync(() -> super.removeLifespan(cacheToUse, key, value, lifespan), blockingExecutor)
+            .thenCompose(Function.identity());
+   }
+
+   @Override
+   CompletableFuture<Boolean> removeMaxIdle(AdvancedCache<K, V> cacheToUse, K key, V value) {
+      // Transactional is still blocking - to be removed later
+      return CompletableFuture.supplyAsync(() -> cacheToUse.removeMaxIdleExpired(key, value), blockingExecutor)
+                  .thenCompose(Function.identity());
+   }
+
+   @Override
+   boolean waitOnLifespanExpiration(boolean hasLock) {
+      // We always have to wait when using optimistic transactions when an entry was found to be expired.
+      // An example is a user calling get then put if the value was null.
+      // This can cause an expiration event to be fired concurrently with the put.
+      // In an optimistic transaction this can cause an issue as the put may get a write skew exception.
```

Comment:
If I understood your arguments correctly, the remove could also succeed after the put and remove the new, non-expired value. Did I understand that right?

Reply:
Yes, if the versions match (when write skew is enabled) or using read committed from my discussion with Pedro.